weatherservices erweitert

hoetzgit opened the pull request:
siehe auch: https://github.com/roe-dl/weewx-DWD/discussions/43

hoetzgit committed:
weatherservices.py customized

TODO: POI Station lat, lon, alt

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
HTTP status code corrected

hoetzgit committed:
minor improvements

- header open-meteo config options docu
- logging

hoetzgit committed:
header open-meteo config documentation

hoetzgit committed:
header service config options documentation

- minor improvements

hoetzgit committed:
debug and logging settings

hoetzgit committed:
dict labels changed

hoetzgit committed:
option to test all open-meteo models

hoetzgit committed:
precipitation, pressure added to open-meteo

- logging

hoetzgit committed:
added 'solar' to default observations CDC, ZAMG

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
HTTP request error evaluation only directly after request

https://github.com/hoetzgit/weewx-DWD/commit/2f10dc9e6c6ec38c46e9b11266fe3b59322c0f3f#comments

hoetzgit committed:
Unnecessary function removed again

see: https://github.com/hoetzgit/weewx-DWD/commit/b8a5b913232c036ba783642daccd4600e040c550#comments

hoetzgit committed:
Use of a specific name for the variable of the station configuration

see: https://github.com/hoetzgit/weewx-DWD/commit/f8bc0d822524e00906dd5705b3c3394c74884331#comments

hoetzgit committed:
Possibility to test all Open-Meteo models

see: https://github.com/hoetzgit/weewx-DWD/commit/f7b5be7a3749b05235cffb639f18bcd638769ba7#comments

hoetzgit committed:
better handling parameter "observations"

see: https://github.com/hoetzgit/weewx-DWD/commit/380a8677a367d4a596bc9dc96e753275df22fd9e#comments

hoetzgit committed:
typo

hoetzgit committed:
Log and error instead of info if a HTTP error occurs

hoetzgit committed:
description of the configuration options

hoetzgit committed:
parameter 'enable' docu corrected

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
initialize data variable

hoetzgit committed:
documentation of the options

hoetzgit committed:
minor changes

hoetzgit committed:
Bright Sky API added

https://brightsky.dev
JSON API for DWD's open weather data.

hoetzgit committed:
renamed class name for current conditions service

hoetzgit committed:
brightsky interval

hoetzgit committed:
brightsky units corrected

hoetzgit committed:
fix unit km/h

hoetzgit committed:
ZAMG units corrected

hoetzgit committed:
DWD CDC 10 Minutes radiation unit

hoetzgit committed:
comment the formula source J/cm^2 to W/m^2

hoetzgit committed:
removed unnessesary converts

hoetzgit committed:
minor improvements

hoetzgit committed:
Convert all value tuple to an appropriate unit in METRIC standardized unit system

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
minor improvements

hoetzgit committed:
fix open-meteo radiation

hoetzgit committed:
minor improvements

hoetzgit committed:
oops

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
If day/night is determined, return value as result

hoetzgit committed:
DWD POI: prevent a WeatherCode from having the value "None"

hoetzgit committed:
Revert "DWD POI: prevent a WeatherCode from having the value "None""

This reverts commit af649b71b5fbe1d9a6349c4114b1bb76bcf9997a.

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Merge branch 'master' into dev

hoetzgit committed:
Aeris "unknown" icon

see: https://github.com/hoetzgit/weewx-DWD/commit/ae85627f158c132e56e39f051b20d71f0c58a139#comments

hoetzgit committed:
Open-Meteo: add new oberservation "precipitation_probability"

hoetzgit committed:
Open-Meteo: Excludelist extended for different models.

Prevents HTTP 400

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1431,183 +2052,580 @@ def getRecord(self):
 
     def __init__(self, engine, config_dict):
-        super(DWDservice,self).__init__(engine, config_dict)
+        super(CurrentService,self).__init__(engine, config_dict)
```
Der Service soll perspektivisch auch die anderen Teile laden. Außerdem muß man genau überlegen, ob man etwas ändert, daß bei einem Update einer bestehenden Installation dazu führt, daß es auf einmal nicht mehr geht. Und das ist die Folge, wenn man DWDservice umbenennt.

hoetzgit replied:
Ich habe das gemacht, weil Du `dwd.py` durch `weatherservices.py` ersetzt hast. Daher auch der "harte Schnitt" in der Konfiguration und in der Benennung des Services. Wobei Deine Hinweise auch zutreffen.

roe-dl replied:
Habe ich das?

Wie auch immer, den "harten Schnitt" können wir uns für später aufheben. Jetzt ist es dafür zu früh.

hoetzgit replied:
> Habe ich das?

Ja ;-)

> Wie auch immer, den "harten Schnitt" können wir uns für später aufheben. Jetzt ist es dafür zu früh.

passe ich an

Da wäre es vorteilhaft gewesen, `dwd.py` nicht zu ersetzen sondern auf `weatherservices.py` zu verweisen. `dwd.py` hätte die bis dahin bestehende Konfiguration genutzt (POI, CDC, ZAMG, Open-Meteo) und `weatherservices.py` konsequent eine  entsprechend angepasste.

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1431,183 +2052,580 @@ def getRecord(self):
-        self.log_failure = weeutil.weeutil.to_bool(site_dict.get('log_failure',True))
-        self.debug = weeutil.weeutil.to_int(site_dict.get('debug',0))
-        if self.debug>0:
+        site_dict = weeutil.config.accumulateLeaves(config_dict.get('WeatherServices',configobj.ConfigObj()))
+        self.log_success = weeutil.weeutil.to_bool(site_dict.get('log_success', True))
+        self.log_failure = weeutil.weeutil.to_bool(site_dict.get('log_failure', True))
+        self.debug = weeutil.weeutil.to_int(site_dict.get('debug', 0))
+        if self.debug > 0:
```
Auch hier ist der Grund für die Abfrage, daß bestehende Installationen bei den Nutzern weiter funktionieren müssen. Das kann man nicht einfach rauslöschen. Dann holt man sich jede Menge Support-Fälle.

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1431,183 +2052,580 @@ def getRecord(self):
         if iconset=='dwd': self.iconset = 5
         if iconset=='aeris': self.iconset = 6
         
-        # deprecated, use section [WeatherServices][[forecast]]
```
Siehe oben.

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1431,183 +2052,580 @@ def getRecord(self):
-                    iconset=station_dict.get('iconset',4),
-                    log_success=weeutil.weeutil.to_bool(station_dict.get('log_success',False)) or self.verbose,
-                    log_failure=weeutil.weeutil.to_bool(station_dict.get('log_failure',True)) or self.verbose)
+                    station_dict,
+                    log_success=weeutil.weeutil.to_bool(station_dict.get('log_success',self.log_success)) or self.verbose,
+                    log_failure=weeutil.weeutil.to_bool(station_dict.get('log_failure',self.log_failure)) or self.verbose)
```
Hatte ich oben schon angemerkt. Ich sehe den Grund für die Zusammenfassung im Moment nicht.

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1431,183 +2052,580 @@ def getRecord(self):
-                    observations=station_dict.get('observations'),
-                    log_success=weeutil.weeutil.to_bool(station_dict.get('log_success',False)) or self.verbose,
-                    log_failure=weeutil.weeutil.to_bool(station_dict.get('log_failure',True)) or self.verbose)
+                    station_dict,
```
Wie oben.

hoetzgit committed:
Unnecessary condition removed, query corrected

see: https://github.com/roe-dl/weewx-DWD/pull/49#pullrequestreview-1361216615
- comment 1, 2

hoetzgit committed:
is_night error case values "N/A" changed to None

see: https://github.com/roe-dl/weewx-DWD/pull/49#pullrequestreview-1361216615
- comment 3

hoetzgit committed:
Condition missing observations corrected

see: https://github.com/roe-dl/weewx-DWD/pull/49#pullrequestreview-1361216615
- comment 12, 16

hoetzgit committed:
Use of the usual notation (comma) in the dictionary definition

see: https://github.com/roe-dl/weewx-DWD/pull/49#pullrequestreview-1361216615
- comment 21

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -217,10 +525,23 @@ def logerr(msg):
+# DS_10 - 10min-sum of diffuse solar radiation - unit J/cm^2
+# GS_10 - 10min-sum of solar incoming radiation - unit J/cm^2
+# LS_10 - 10min-sum of longwave downward radiation - unit J/cm^2
+if weewx.units.conversionDict.get('joule_per_cm_squared_10minutes') is None:
+    weewx.units.conversionDict.setdefault('joule_per_cm_squared_10minutes',{})
```
passe ich an

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -217,10 +525,23 @@ def logerr(msg):
+# LS_10 - 10min-sum of longwave downward radiation - unit J/cm^2
+if weewx.units.conversionDict.get('joule_per_cm_squared_10minutes') is None:
+    weewx.units.conversionDict.setdefault('joule_per_cm_squared_10minutes',{})
+if weewx.units.conversionDict['joule_per_cm_squared_10minutes'].get('watt_per_meter_squared') is None:
```
passe ich an

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -253,13 +574,83 @@ def wget(url,log_success=False,log_failure=True):
+    except LookupError as e:
+        if log_failure:
+            logerr("thread '%s': is_night %s - %s" % (self.name, e.__class__.__name__, e))
+        return "N/A"
```
passe ich an

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -370,25 +776,39 @@ class DWDPOIthread(BaseThread):
     
-    def __init__(self, name, location, prefix, iconset=4, log_success=False, log_failure=True):
-    
+    def __init__(self, name, poi_dict, log_success=False, log_failure=True):
+
```
Kompakter Funktionskopf, da viele Parameter aus dem Dictionary genutzt werden.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -370,25 +776,39 @@ class DWDPOIthread(BaseThread):
+
+        self.station = poi_dict.get('station')
+        self.iconset = weeutil.weeutil.to_int(poi_dict.get('iconset', 4))
+        self.lat = weeutil.weeutil.to_float(poi_dict.get('latitude'))
+        self.lon = weeutil.weeutil.to_float(poi_dict.get('longitude'))
+        self.alt = weeutil.weeutil.to_float(poi_dict.get('altitude'))
```
Der POI Teil ist nicht weiter verändert. Die Koordinaten sind im Ergebnis nicht enthalten, bzw. ich habe es nicht gesehen.  Option die Koordinaten der Station mit anzugeben.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -370,25 +776,39 @@ class DWDPOIthread(BaseThread):
             else:
                 obsgroup = weewx.units.obs_group_dict.get(obstype)
             if obsgroup:
+                weewx.units.obs_group_dict.setdefault(obstype, obsgroup)
```
Hier reicht mein Wissen nicht. Was passiert im ungünstigsten Fall? Wird hier nicht nur eine Gruppe zugeordnet, wenn diese Zuordnung nicht existiert, sonst nur Rückgabe?

roe-dl replied:
Der Nutzer kann `obstype` (also den Namen des Observation types) ohne prefix anderweitig genutzt haben. Mit einer anderen Gruppe. Warum sollten wir einen observation type definieren, den wir gar nicht benutzen?

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -451,7 +871,7 @@ def get_ww(present_weather, night):
     def getRecord(self):
         """ download and process POI weather data """
-        url = 'https://opendata.dwd.de/weather/weather_reports/poi/'+self.location+'-BEOB.csv'
+        url = 'https://opendata.dwd.de/weather/weather_reports/poi/'+self.station+'-BEOB.csv'
```
> Eine Station ist das nämlich nicht immer.

Was ist es dann?

roe-dl replied:
POI steht für point of interest, also touristisch interessanter Ort. Da scheint mir das Wort location deutschlich besser zu passen.

hoetzgit replied:
Können wir uns darauf einigen, dass es sich um Messwerte einer DWD Station handelt? Du verwendest z.B. in der kompletten Dokumentation dieser Funktionalität den Begriff "Station".

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -487,6 +907,8 @@ def getRecord(self):
                         y['time'] = (val,None,None)
                     else:
                         col = DWDPOIthread.OBS.get(names[idx])
+                        if col is None:
+                            continue
```
Ich glaube mich zu erinnern, ich hatte später eine fehlerhafte Zuordnung, da `col` (bisher) nur in Verbindung mit `val` geprüft wurde.

roe-dl replied:
Aha.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -534,12 +982,12 @@ class DWDCDCthread(BaseThread):
         'RWS_DAU_10':('rainDur','minute','group_deltatime'),
         'RWS_10':('rain','mm','group_rain'),
-        'RWS_IND_10':('rainIndex','',''),
+        'RWS_IND_10':('rainIndex',None,None),
```
Also wenn ich es richtig verstanden habe, gilt `None` für einen nicht spezifizierten Unit Typ. Ich denke es ist besser als ein leerer String. Meintest Du das?

roe-dl replied:
Die Konvertierungsfunktionen von WeeWX müssen damit umgehen können. Ich habe nicht geprüft, ob die mit `None` auch zurechtkommen. Das wäre zu prüfen. Was steht in `units.py`?

hoetzgit replied:
Die kommen damit zurecht und haben eher Probleme mit einem leeren String.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -534,12 +982,12 @@ class DWDCDCthread(BaseThread):
         # solar
-        'DS_10':('solarRad','J/cm^2','group_radiation'),
-        'GS_10':('radiation','J/cm^2','group_radiation'),
+        'DS_10':('solarRad','joule_per_cm_squared_10minutes','group_radiation'),
+        'GS_10':('radiation','joule_per_cm_squared_10minutes','group_radiation'),
```
Auch da reicht mein Verständnis nicht. Ich denke, es müsste tatsächlich die Einheit ohne `10minutes` bezeichnet werden und dann in die Zieleinheit konvertiert werden.

hoetzgit replied:
Was könnten wir hier machen? Da bin ich überfordert, hast Du eine Idee?

roe-dl replied:
Wir nehmen die Änderung hier erst einmal wieder raus und denken später darüber nach.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -548,39 +996,50 @@ class DWDCDCthread(BaseThread):
-            observations = ('air','wind','gust','precipitation')
+
+        observations = cdc_dict.get('observations')
+        if observations is not None:
```
passe ich an

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -828,8 +1303,11 @@ def __init__(self, name, location, prefix, iconset=4, observations=None, user=''
-        if not observations:
-            observations = ('air','wind','gust','precipitation')
+        observations = zamg_dict.get('observations')
+        if observations is not None:
```
passe ich an

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -964,68 +1446,104 @@ def getRecord(self):
+        ,'windgusts_10m': 'windGust' # not available in forecast model ecmwf
+        ,'cloudcover': 'cloudcover'
+        ,'evapotranspiration': 'et'
+        ,'precipitation': 'precipitation'
```
Meinst Du `precipitation`? Laut Doku: "Total precipitation (rain, showers, snow) sum of the preceding hour"

roe-dl replied:
Ok.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -964,68 +1446,104 @@ def getRecord(self):
+        ,'windspeed_10m': 'windSpeed'
+        ,'windgusts_10m': 'windGust' # not available in forecast model ecmwf
+        ,'cloudcover': 'cloudcover'
+        ,'evapotranspiration': 'et'
```
Das hatte ich bewusst so erstellt und auch kommentiert. Lässt ein Benutzer das Präfix weg, würde durch die Umwandlung `ET` entstehen. So entsteht `Et` und überschreibt nicht (evtl. unbeabsichtigt) das Feld `ET`. Möchte man trotzdem `ET` kann man dies durch Mapping erreichen. Gleiches bei `UV`

roe-dl replied:
Ah, verstanden.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -964,68 +1446,104 @@ def getRecord(self):
+        'temperature': u'°C'
+        ,'windspeed': 'km/h'
+        ,'winddirection': u'°'
+        ,'weathercode': 'wmo code'
```
`weathercode`? Nur eine "Dummy" Einheit für den WMO Code von Open-Meteo

roe-dl replied:
Ich meine, ist das Code-Tabell 4677 oder 4680? Kannst Du das in einen Kommentar schreiben?

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -1083,52 +1601,146 @@ class DWDOPENMETEOthread(BaseThread):
-        for opsapi, obsweewx in DWDOPENMETEOthread.CURRENTOBS.items():
+
+        station = openmeteo_dict.get('station', 'ThisStation')
+        self.lat = weeutil.weeutil.to_float(openmeteo_dict.get('latitude'))
+        self.lon = weeutil.weeutil.to_float(openmeteo_dict.get('longitude'))
+        self.alt = weeutil.weeutil.to_float(openmeteo_dict.get('altitude'))
```
Die Eintragungen kommen aus der Konfig. Entweder eigene Station oder eine andere Station. Meintest Du das?

roe-dl replied:
Die Koordinaten der eigenen Station sollten aus `stn_info` kommen.

hoetzgit replied:
Kommen sie, wenn z.B. "ThisStation" verwendet wird. Sonst aus der Parametrierung der entsprechenden Station.

hoetzgit commented on bin/user/weatherservices.py:
```diff
@@ -1168,21 +1780,26 @@ def get_ww(wwcode, night):
 
-        # DWD API endpoint "v1/dwd-icon"
-        baseurl = 'https://api.open-meteo.com/v1/'+self.model
+        baseurl = 'https://api.open-meteo.com/v1/%s' % endpoint
```
Weil ich zusätzlich zu Endpoint `v1/dwd-icon` auch `v1/forecast` nutzen möchte

hoetzgit commented:
Danke für Deine Hinweise und Kommentare, ein paar Rückfragen habe ich ergänzt. Die Anpassungen versuche ich möglichst bald zu erledigen, zeitlich bei mir auch etwas schlecht z.Zt.

hoetzgit committed:
unnecessary function removed

hoetzgit committed:
rename service undo

hoetzgit committed:
backward compatibility restored

hoetzgit committed:
DWD POI: station renamed to location

hoetzgit committed:
backward compatibility restored

station is now in station_dict

hoetzgit committed:
bugfix is_night logerr

hoetzgit committed:
Support of "Brightsky API" temporarily removed

Currently, the Brightsky API is being further developed. After completion, the Brightsky API could be reactivated.

hoetzgit committed:
determination day/night only if lat and lon known

hoetzgit committed:
typo

hoetzgit committed:
set default station only if lat or lon is None

hoetzgit committed:
weewx almanac altitude is optional

roe-dl commented on bin/user/weatherservices.py:
```diff
@@ -1655,6 +2221,9 @@ def _to_weewx(self, thread_name, reply, usUnits):
             else:
                 try:
                     val = reply[key]
+                    # first convert val to standard unit in METRIC unit system
+                    val = weewx.units.convertStd(val, weewx.METRIC)
+                    # now convert val to archive record unit system
```
Warum das? Der konvertiert von jeder zu jeder Einheit. Wenn man Einheiten anlegt, muß man übrigens grundsätzlich auch alle Umrechnungen zwischen der neuen Einheit und den vorhandenen Einheiten eintragen. Man kann nie vorhersehen, wo irgendetwas hin- und herkonvertiert wird.

hoetzgit replied:
Ich hatte ein Problem, dass die metrische Einheit nicht in die US Einheit konvertiert werden konnte, da es keine Konvertierungsfunktion gab. Ich weiß nicht mehr genau was es war, evtl. 'km' zu 'foot' oder so. Daher habe ich erst die Konvertierung in die Standardeinheit durchgeführt und dann zum Archiv-Record Unit System. Mir fehlt da einfach zu viel um das vernünftig zu machen. Du hast jetzt mehr Arbeit das durchzusehen als wenn Du es gleich richtig gemacht hättest.

roe-dl commented:
Wollen wir das noch zu Ende bringen? Ich war jetzt mit ganz anderem beschäftigt, aber jetzt scheint es Nachfragen zu geben.

hoetzgit commented:
Hallo Johanna, danke der Nachfrage. Aktuell bastel ich mir selbst etwas auf Basis weatherservices.py. Da kann ich Fehler machen und lernen. Ich denke, für die Allgemeinheit ist mein Kenntnisstand zu gering. Daher bin ich leider raus. Wenn ich sehe, was Du in letzter Zeit so gemacht hast, dann ist Deine Lernkurve eindeutig steiler als meine.
Viele Grüsse Henry

roe-dl commented:
Das ist schade.

Das Distrometer, wenn Du das meinst, habe ich eigentlich angeschafft, weil es mich nervte, daß das Symbol immer wieder Regen zeigte, obwohl es nicht regnete oder umgekehrt. Die ganze Geschichte hat sich als viel komplizierter erwiesen als gedacht. Das Gerät liefert bei weitem nicht alle Wetter-Codes, und es irrt sich relativ häufig. Das ist alles im Einklang mit der Spezifikation, erfordert aber eben einiges an Nachbereitung. Das hatte ich so nicht erwartet.
